internal/oci/runtime_oci.go

```diff
@@ -898,73 +899,89 @@ func (r *runtimeOCI) AttachContainer(c *Container, inputStream io.Reader, output
 	return nil
 }
 
-// PortForwardContainer forwards the specified port provides statistics of a container.
-func (r *runtimeOCI) PortForwardContainer(c *Container, port int32, stream io.ReadWriter) error {
-	emptyStreamOnError := true
-	defer func() {
-		if emptyStreamOnError {
-			go func() {
-				if _, copyError := pools.Copy(ioutil.Discard, stream); copyError != nil {
-					logrus.Errorf("error closing port forward stream after other error: %v", copyError)
-				}
-			}()
+// PortForwardContainer forwards the specified port into the provided container.
+func (r *runtimeOCI) PortForwardContainer(
+	ctx context.Context, c *Container, nsPath string,
+	port int32, stream io.ReadWriteCloser,
+) error {
+	log.Infof(ctx,
+		"Starting port forward for %s in network namespace %s", c.ID(), nsPath,
+	)
+
+	// Adapted reference implementation:
+	// https://github.com/containerd/cri/blob/8c366d/pkg/server/sandbox_portforward_unix.go#L65-L120
+	if err := ns.WithNetNSPath(nsPath, func(_ ns.NetNS) error {
+		defer stream.Close()
+
+		// TODO: hardcoded to tcp4 because localhost resolves to ::1 by default
+		// if the system has IPv6 enabled. Theoretically happy eyeballs will
+		// try IPv6 first and fallback to IPv4 but resolving localhost doesn't
+		// seem to return and IPv4 address, thus failing the connection.
+		conn, err := net.Dial("tcp4", fmt.Sprintf("localhost:%d", port))
+		if err != nil {
+			return errors.Wrapf(err, "dialing %d", port)
 		}
-	}()
-	containerPid := c.State().Pid
-	socatPath, lookupErr := exec.LookPath("socat")
-	if lookupErr != nil {
-		return fmt.Errorf("unable to do port forwarding: socat not found")
-	}
+		defer conn.Close()
 
-	args := []string{"-t", fmt.Sprintf("%d", containerPid), "-n", socatPath, "-", fmt.Sprintf("TCP4:localhost:%d", port)}
+		errCh := make(chan error, 2)
 
-	nsenterPath, lookupErr := exec.LookPath("nsenter")
-	if lookupErr != nil {
-		return fmt.Errorf("unable to do port forwarding: nsenter not found")
-	}
+		debug := func(format string, args ...interface{}) {
+			log.Warnf(ctx, fmt.Sprintf(
```

Comment:
it strikes me a bit odd the function is called debug but it uses warn level

Reply:
Ah this should not happen, reverting it.